src/core/ext/filters/client_channel/resolver/polling_resolver.cc

```diff
@@ -0,0 +1,201 @@
+                                 TraceFlag* tracer)
+    : authority_(args.uri.authority()),
+      name_to_resolve_(absl::StripPrefix(args.uri.path(), "/")),
+      channel_args_(grpc_channel_args_copy(channel_args)),
+      work_serializer_(std::move(args.work_serializer)),
+      result_handler_(std::move(args.result_handler)),
+      tracer_(tracer),
+      interested_parties_(args.pollset_set),
+      min_time_between_resolutions_(min_time_between_resolutions),
+      backoff_(backoff_options) {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] created", this);
+  }
+}
+
+PollingResolver::~PollingResolver() {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] destroying", this);
+  }
+  grpc_channel_args_destroy(channel_args_);
+}
+
+void PollingResolver::StartLocked() { MaybeStartResolvingLocked(); }
+
+void PollingResolver::RequestReresolutionLocked() {
+  if (request_ == nullptr) {
+    MaybeStartResolvingLocked();
+  }
+}
+
+void PollingResolver::ResetBackoffLocked() {
+  if (have_next_resolution_timer_) {
+    grpc_timer_cancel(&next_resolution_timer_);
+  }
+  backoff_.Reset();
+}
+
+void PollingResolver::ShutdownLocked() {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] shutting down", this);
+  }
+  shutdown_ = true;
+  if (have_next_resolution_timer_) {
+    grpc_timer_cancel(&next_resolution_timer_);
+  }
+  request_.reset();
+}
+
+void PollingResolver::OnNextResolution(void* arg, grpc_error_handle error) {
+  auto* self = static_cast<PollingResolver*>(arg);
+  (void)GRPC_ERROR_REF(error);  // ref owned by lambda
+  self->work_serializer_->Run(
+      [self, error]() { self->OnNextResolutionLocked(error); }, DEBUG_LOCATION);
+}
+
+void PollingResolver::OnNextResolutionLocked(grpc_error_handle error) {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO,
+            "[polling resolver %p] re-resolution timer fired: error=\"%s\", "
+            "request_=%p",
+            this, grpc_error_std_string(error).c_str(), request_.get());
+  }
+  have_next_resolution_timer_ = false;
+  if (error == GRPC_ERROR_NONE && request_ == nullptr) {
+    StartResolvingLocked();
+  }
+  Unref(DEBUG_LOCATION, "retry-timer");
+  GRPC_ERROR_UNREF(error);
+}
+
+void PollingResolver::OnRequestComplete(Result result) {
+  Ref(DEBUG_LOCATION, "OnRequestComplete").release();
+  work_serializer_->Run(
+      [this, result]() mutable { OnRequestCompleteLocked(std::move(result)); },
+      DEBUG_LOCATION);
+}
+
+void PollingResolver::OnRequestCompleteLocked(Result result) {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] request complete", this);
+  }
+  request_.reset();
+  if (!shutdown_) {
+    if (result.service_config.ok() && result.addresses.ok()) {
```

Comment:
shouldn't this be `result.service_config.ok() || result.addresses.ok()` ?

Otherwise, this looks like a behavioral change from the c-ares resolver current implementation, which resets backoff in the case that we resolved only addresses, without a service config

Reply:
Related: I think `result.service_config` will typically be at the default `unknown` value, since the c-ares resolver doesn't request service config by default and so won't fill it in

Reply:
The `service_config` field defaults to a null value (i.e., OK status):

https://github.com/grpc/grpc/blob/fe913387da494713a945f7842669bd3bf3c27a03/src/core/lib/resolver/resolver.h#L57

This means that `result.service_config.ok()` will be false only if the resolver explicitly returns an error for it.  I think if we get an explicit error for either service config or addresses, we should consider it a failure for the purposes of backoff.

I don't think this will actually change the behavior of the c-ares resolver, because it does not explicitly return an error for the service config if it is not fetching the service config.

Reply:
Ah, got it, I misread that that it defaults to a null value